### DISCUSSION TITLE
Python, ~/.kube/config, Configuration and load_kube_config

### DISCUSSION THREAD

**Discussion by bensternlieb** [argoproj/argo-workflows#8011] [Q&A] [answered]
Folks,

We run on eks and typically had code that followed this pattern when using the older argo-workflows==3.5.1 API:
```
from argo.workflows.config import load_kube_config
...

    def submit(self):
        """Generates and submits a cron workflow."""
        load_kube_config()
        V1alpha1Api().create_namespaced_workflow(
            "default", self.get_cron_workflow()
        )
 ```

We  would like to upgrade to argo-workflows-5.0.0, but the `load_kube_config` pattern doesn't seem to be supported anymore - example code only shows explicit configuration via URL.  What is the analogous way to used ~/.kube/config in 5.0.0?

**Comment by terrytangyuan** [comment, marked as answer]
Currently, we only support submitting workflow through Argo server. We could potentially support this though. 

Any thoughts? @flaviuvadan

**Reply by bensternlieb** [reply]
I've just spent some time in the `argo-workflows  6.3.0` API - really miss the ability to easily generate yaml and/or use pure kubernetes API.  So this is a bit of a deal-breaker for our team. We frankly never use or need a GUI for our jobs.  

With `argo-workflows  3.5.1` all we needed was EKS access to push jobs - we neither had to port forward or ingress-expose the argo-server, and didn't have to worry about security constructs or additional endpoint review from our Security team.

I believe the glory of argo was it's Kubernetes Custom Resource integration.  The move to a Separate Web API must have some justification, but it's a complexity bummer for those who have enjoyed using kubernetes and argo apis together seamlessly.

**Reply by flaviuvadan** [reply]
@bensternlieb is it possible to submit the workflow YAML file via `kubectl apply -f my_workflow.yaml` if the workflow spec file is generated with the Python SDK? The workflow file can be serialized to a YAML and submitted directly, which does use the credentials from `~./kube/config`.

**Comment by flaviuvadan** [comment]
@bensternlieb there's a Hera example with `argo-workflows` v6 that generates a token for a specific K8S SA and submits a workflow: https://github.com/argoproj-labs/hera-workflows/blob/main/examples/k8s_sa.py Alternatively, [port forwarding](https://github.com/argoproj-labs/hera-workflows/blob/main/examples/k8s_sa.py) relies on `kubectl` to use the config file, and then workflows can be submitted to `localhost:2746`.

**Reply by nate-selzer** [reply]
Hey @flaviuvadan thanks for the response here. I'm picking up the client upgrade task for our team, and am wondering about the best path forward. I agree with @bensternlieb that using the k8s api instead of custom tooling for argo is nice, however, if this statement
> Currently, we only support submitting workflow through Argo server

is still true, I think we'd want to go with whatever the supported route is.

My questions:

1. Is it still true that using the k8s api to create/modify workflows is not supported?
2. Do you have a summary of what we're risking/missing out on if we choose to stay on the unsupported route (using k8s api)? As of right now it still seems like all argo resources are defined as k8s CRDs, so using the argo-cli/sdk to generate yaml and then using the k8s api should "just work". Is this correct, and do you think it will be correct for the foreseeable future?
3. In the argo docs, I'm seeing a number of places where argo resources are modified using `kubectl`. An [example](https://argoproj.github.io/argo-workflows/walk-through/artifacts/#what-happens-if-garbage-collection-fails). Is there a rundown of when it's appropriate to use the k8s api to modify argo resources vs. when we should be using the argo api directly?

**Reply by flaviuvadan** [reply]
You can now use Hera to generate a YAML file, which you can submit via `kubectl apply -f`!
